[Inference Providers] How to register as a provider (#1663)
huggingface/hub-docs: docs/inference-providers/register-as-a-provider.md

# How to be registered as an inference provider on the Hub?

<Tip>

> Want to be an Inference Provider on the Hub? Please reach out to us!

Reviewer: ```suggestion
Want to be an Inference Provider on the Hub? Please reach out to us on social networks or [here on the Hub](https://huggingface.co/spaces/huggingface/HuggingDiscussions/discussions/49)
```

</Tip>

This guide details the steps for registering as an inference provider on the Hub and provides implementation guidance.

1. **Implement standard task APIs** - Follow our task API schemas for compatibility (see [Prerequisites](#1-prerequisites)).
2. **Submit a PR for JS client integration** - Add your provider to [huggingface.js](https://github.com/huggingface/huggingface.js/tree/main/packages/inference) (see [JS Client Integration](#2-js-client-integration)).
3. **Register model mappings** - Use our Model Mapping API to link your models to Hub models (see [Model Mapping API](#3-model-mapping-api)).
4. **Implement a billing endpoint** - Provide an API for billing (see [Billing](#4-billing)).
5. **Submit a PR for Python client integration** - Add your provider to [huggingface_hub](https://github.com/huggingface/huggingface_hub) (see [Python client integration](#5-python-client-integration)).
6. **Provide an icon** - Submit an SVG icon for your provider.

> 7. **Create documentation** - Add documentation and do some communication on your side.

Reviewer: ```suggestion
7. **Create documentation** - Add documentation and do a lot of communication on your side.
```

> 8. **Add a documentation page** - Add a provider-specific page in the Hub documentation.

Reviewer: ```suggestion
8. **Add a documentation page** - Add a provider-specific page in the Hub documentation.
```

> 8. **Add a documentation page** - Add a provider-specific page in the Hub documentation.

Reviewer: ```suggestion
8. **Add a documentation page here** - Add a provider-specific page in the Hub documentation.
9. **Share share share** do a lot of comms so that your integration is as successful as possible!
```

If the provider supports multiple tasks that require different implementations, create dedicated subclasses for each task, following the pattern used in the existing providers implementation, e.g. [Together AI provider implementation](https://github.com/huggingface/huggingface.js/blob/main/packages/inference/src/providers/together.ts).

For text-generation and conversational tasks, you can just inherit from BaseTextGenerationTask and BaseConversationalTask respectively (defined in [providerHelper.ts]((https://github.com/huggingface/huggingface.js/blob/main/packages/inference/src/providers/providerHelper.ts))) and override the methods if needed. Examples can be found in [Cerebras](https://github.com/huggingface/huggingface.js/blob/main/packages/inference/src/providers/cerebras.ts) or [Fireworks](https://github.com/huggingface/huggingface.js/blob/main/packages/inference/src/providers/fireworks.ts) provider implementations.

### Register the provider

> Go to [packages/inference/src/lib/getProviderHelper.ts](https://github.com/huggingface/huggingface.js//blob/main/packages/inference/src/lib/getProviderHelper.ts) and add your provider to `PROVIDERS`. Please try to respect alphabetical order. 

Reviewer 2: An additional step here is to get in touch with us to add the provider to moon-landing and map to the organization

Author: we added a Tip in the Model Mapping API section for that: https://github.com/huggingface/hub-docs/blob/549ecd03e9f5b818e45126f20115d0e95e79a0eb/docs/inference-providers/register-as-a-provider.md?plain=1#L130

but I'm okay to add a specific section for that, wdyt?

Author: I've added in the introduction a "Reach out to us to add your provider server-side and provide your SVG icon" step just after the JS integration https://github.com/huggingface/hub-docs/pull/1663/commits/a179888a33ea8858aa982c49172203865cd05f6f

> ## 3. Model Mapping API

Author: where should we put the OpenAPI specs yaml files?

> ## 3. Model Mapping API

Reviewer: at some point but maybe not right now:)

<Tip>

To proceed with this step, we have to enable your account server-side. Make sure you have an organization on the Hub for your enterprise.

</Tip>

### Register a mapping item

```http
POST /api/partners/{provider}/models
```
Create a new mapping item, with the following body (JSON-encoded):

```json
{
    "task": "WidgetType", // required
    "hfModel": "string", // required: the name of the model on HF: namespace/model-name
    "providerModel": "string", // required: the partner's "model id" i.e. id on your side
    "status": "live" | "staging" // Optional: defaults to "staging". "staging" models are only available to members of the partner's org, then you switch them to "live" when they're ready to go live
}
```

- `task`, also known as `pipeline_tag` in the HF ecosystem, is the type of model / type of API
(examples: "text-to-image", "text-generation", but you should use "conversational" for chat models)
- `hfModel` is the model id on the Hub's side.
- `providerModel` is the model id on your side (can be the same or different).

In the future, we will add support for a new parameter (ping us if it's important to you now):
```json
{

>     "hfFilter": ["string"] // (both can't be defined at the same time)

Reviewer 2: ```suggestion
    "hfFilter": ["string"]
```

# How to be registered as an inference provider on the Hub?

<Tip>

> Want to be an Inference Provider on the Hub? Please reach out to us on social networks or [here on the Hub](https://huggingface.co/spaces/huggingface/HuggingDiscussions/discussions/49).

Reviewer 3: ```suggestion
Want to be listed as an Inference Provider on the Hugging Face Hub? Please reach out to us on social networks or [here on the Hub](https://huggingface.co/spaces/huggingface/HuggingDiscussions/discussions/49).
```

</Tip>

This guide details the steps for registering as an inference provider on the Hub and provides implementation guidance.

1. **Implement standard task APIs** - Follow our task API schemas for compatibility (see [Prerequisites](#1-prerequisites)).
2. **Submit a PR for JS client integration** - Add your provider to [huggingface.js](https://github.com/huggingface/huggingface.js/tree/main/packages/inference) (see [JS Client Integration](#2-js-client-integration)).
3. **Register model mappings** - Use our Model Mapping API to link your models to Hub models (see [Model Mapping API](#3-model-mapping-api)).
4. **Implement a billing endpoint** - Provide an API for billing (see [Billing](#4-billing)).
5. **Submit a PR for Python client integration** - Add your provider to [huggingface_hub](https://github.com/huggingface/huggingface_hub) (see [Python client integration](#5-python-client-integration)).

> 6. **Provide an icon** - Submit an SVG icon for your provider.

Reviewer 3: where?

Reviewer 3: this maybe we can remove and do it be ourselves as most of the time their own icon is quite bad cc @ChunTeLee

Reviewer 4: (I do think that at some point the SVGs should be in inference.js instead of moon-landing but for now agree with your suggestion)

7. **Create documentation on your side** - Add documentation and do a lot of communication on your side.

> 8. **Add a documentation page here** - Add a provider-specific page in the Hub documentation.

Reviewer 3: ```suggestion
8. **Add a documentation page** - Open a Pull Request in this repo (huggingface/hub-docs) to add a provider-specific page in the documentation.
```

implementation downstream (for Python SDK, to generate code snippets, etc.).

### What is a Task 

You will see that inference methods (`textToImage`, `chatCompletion`, etc.) have names that closely
mirror the task names. A task, also known as `pipeline_tag` in the HF ecosystem, is the type of
model (basically which types of inputs and outputs the model has), for instance "text-generation"
or "text-to-image". It is indicated prominently on model pages, here:

<div class="flex justify-center">
    <picture>
        <img class="block dark:hidden" src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/inference-providers/pipeline-tag-on-model-page-light.png">
        <img class="hidden dark:block" src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/inference-providers/pipeline-tag-on-model-page-dark.png">
    </picture>
</div>

The list of all possible tasks can be found at https://huggingface.co/tasks and the list of JS method names is documented in the README at https://github.com/huggingface/huggingface.js/tree/main/packages/inference.

<Tip>

> Note that `chatCompletion` is an exception as it is not a pipeline_tag, per se. Instead, it

Reviewer 3: nice 👍

    }

	makeRoute(params: UrlParams): string {
        // Return the route to use for the request. e.g. /v1/chat/completions route is commonly use for chat completion.
		throw new Error("Needs to be implemented");
	}

	preparePayload(params: BodyParams): Record<string, unknown> {
        // Return the payload to use for the request, as a dict.
		throw new Error("Needs to be implemented");
	}

	getResponse(response: unknown, outputType?: "url" | "blob"): string | Promise<Blob>{
		// Return the response in the expected format.
        throw new Error("Needs to be implemented");
    }
}
```

Implement the methods that require custom handling. Check out the base implementation to check default behavior. If you don't need to override a method, just remove it. You have to define at least `makeRoute`, `preparePayload` and `getResponse`.

If the provider supports multiple tasks that require different implementations, create dedicated subclasses for each task, following the pattern used in the existing providers implementation, e.g. [Together AI provider implementation](https://github.com/huggingface/huggingface.js/blob/main/packages/inference/src/providers/together.ts).

> For text-generation and conversational tasks, you can just inherit from BaseTextGenerationTask and BaseConversationalTask respectively (defined in [providerHelper.ts]((https://github.com/huggingface/huggingface.js/blob/main/packages/inference/src/providers/providerHelper.ts))) and override the methods if needed. Examples can be found in [Cerebras](https://github.com/huggingface/huggingface.js/blob/main/packages/inference/src/providers/cerebras.ts) or [Fireworks](https://github.com/huggingface/huggingface.js/blob/main/packages/inference/src/providers/fireworks.ts) provider implementations.

Reviewer 4: ```suggestion
For text-generation and conversational tasks, you can just inherit from `BaseTextGenerationTask` and `BaseConversationalTask` respectively (defined in [providerHelper.ts]((https://github.com/huggingface/huggingface.js/blob/main/packages/inference/src/providers/providerHelper.ts))) and override the methods if needed. Examples can be found in [Cerebras](https://github.com/huggingface/huggingface.js/blob/main/packages/inference/src/providers/cerebras.ts) or [Fireworks](https://github.com/huggingface/huggingface.js/blob/main/packages/inference/src/providers/fireworks.ts) provider implementations.
```

        { "requestId": "deadbeef3", "costNanoUsd": 100 }
    ]
}
```

### Price Unit

We require the price to be an **integer** number of **nano-USDs** (10^-9 USD).

### How to define the request ID

For each request/generation you serve, you should define a unique request (or response) ID,
and provide it as a response Header. We will use this ID as the request ID for the billing API
above.

As part of those requirements, please let us know your Header name. If you don't already have one, we suggest the `Inference-Id` name for instance, and it should contain a UUID character string.

**Example**: Defining an `Inference-Id` header in your inference response.

```http
POST /v1/chat/completions
Content-Type: application/json
[request headers]
[request body]
------
HTTP/1.1 200 OK
Content-Type: application/json
[other request headers]
Inference-Id: unique-id-00131
[response body]
```

## 5. Python client integration

<Tip>

Before adding a new provider to the `huggingface_hub` Python library, make sure that all the previous steps have been completed and everything is working on the Hub. Support in the Python library comes as a second step. 

</Tip>

> ### Implement the provider helper

Reviewer 4: ```suggestion
### Implement the provider helper (Python)
```

(see above)

# How to be registered as an inference provider on the Hub?

> If you'd like to be an inference provider on the Hub, you must follow the steps outlined in this guide.

Reviewer 4: I would briefly mention in the introduction what we expect from them:
- Use tasks
- PR for JS client 
- Mapping
- Billing endpoint
- PR for Python client

with links to the corresponding docs

Reviewer 4: also: provide an icon (svg)

Reviewer 4: + add documentation on their side
+ make comms!
+ as last step: add documentation page in HF docs (not done yet but we'll have 1 doc per provider)

Author: addressed [here](https://github.com/huggingface/hub-docs/pull/1663/files/6e3836097655b77f3118410d33c78888e348051b..40e7e85253a1efc73a0ddadaf6465186da16b664).

> ## 1. Requirements

Reviewer 4: ```suggestion
## 1. Prerequisites
```
(nit :see_no_evil:)

> <Tip>
> 
> If you provide inference only for LLMs and VLMs following the OpenAI API, you
> can probably skip most of this section and just open a PR on
> https://github.com/huggingface/huggingface.js/tree/main/packages/inference to add you as a provider.
> 
> </Tip>

Reviewer 4: ```suggestion
<Tip>

If your implementation strictly follows the OpenAI API for LLMs and VLMs, you may be able to skip most of this section. In that case, simply open a PR on [huggingface.js](https://github.com/huggingface/huggingface.js/tree/main/packages/inference) to register.

</Tip>
```

> The first step to understand the integration is to take a look at the JS inference client that lives
> inside the huggingface.js repo:
> https://github.com/huggingface/huggingface.js/tree/main/packages/inference

Reviewer 4: ```suggestion
The first step to understand the integration is to take a look at the JS inference client that lives inside the huggingface.js repo: https://github.com/huggingface/huggingface.js/tree/main/packages/inference
```

(I usually tend to prefer 1 paragraph == 1 code line => easier to review when making changes to it afterwards)

(same for the whole module)

> This is the client that powers our Inference widgets on model pages, and is the blueprint
> implementation for other downstream SDKs like Python's `huggingface-hub` and other tools.

Reviewer 4: ```suggestion
This is the client that powers our Inference widgets on model pages, and is the blueprint
implementation downstream (for Python SDK, to generate code snippets, etc.).
```

This is not an issue for LLMs as everyone converged on the OpenAI API anyways, but can be
more tricky for other tasks like "text-to-image" or "automatic-speech-recognition" where there
exists no standard API.

For example, you can find the expected schema for Text to Speech here: [https://github.com/huggingface/huggingface.js/packages/src/tasks/text-to-speech/spec/input.json#L4](https://github.com/huggingface/huggingface.js/blob/0a690a14d52041a872dc103846225603599f4a33/packages/tasks/src/tasks/text-to-speech/spec/input.json#L4), and similarly for other supported tasks. If your API for a given task is different from HF's, it is not an issue: you can tweak the code in `huggingface.js` to be able to call your models, i.e., provide some kind of "translation" of parameter names and output names. However, API specs should not be model-specific, only task-specific. Run the JS code and add some [tests](https://github.com/huggingface/huggingface.js/blob/main/packages/inference/test/HfInference.spec.ts) to make sure it works well. We can help with this step!

> ## 2. JS Client Integration

Author: this section is based on the refacto implemented here https://github.com/huggingface/huggingface.js/pull/1315 (not merged yet)

Before proceeding with the next steps, ensure you've implemented the necessary code to integrate with the JS client and thoroughly tested your implementation. Here are the steps to follow:

> ### 1. Implement the provider helper

Reviewer 4: ```suggestion
### Implement the provider helper
```

I would keep the numbered items only for the `##` section titles. Having 2 levels of numbers is confusing IMO

(same for other section titles)

> Create a new file under packages/inference/src/providers/{provider_name}.ts and copy-paste the following snippet.

Reviewer 4: ```suggestion
Create a new file under `./packages/inference/src/providers/{provider_name}.ts` and copy-paste the following snippet.
```

```ts
import { TaskProviderHelper } from "./providerHelper";

export class MyNewProviderTask extends TaskProviderHelper {

	constructor() {
		super("your-provider-name", "your-api-base-url", "task-name");
	}

    override prepareHeaders(params: HeaderParams, binary: boolean): Record<string, string> {
        // Override the headers to use for the request.
        return super.prepareHeaders(params, binary);
    }

	makeRoute(params: UrlParams): string {
        // Return the route to use for the request. e.g. /v1/chat/completions route is commonly use for chat completion.
		throw new Error("Needs to be implemented");
	}

	preparePayload(params: BodyParams): Record<string, unknown> {
        // Return the payload to use for the request, as a dict.
		throw new Error("Needs to be implemented");
	}

> 	getResponse(response: TogetherBase64ImageGeneration, outputType?: "url" | "blob"): string | Promise<Blob>{

Reviewer 4: ```suggestion
	getResponse(response: <something else?>, outputType?: "url" | "blob"): string | Promise<Blob>{
```

Author: 🙈 wrong copy-paste

		// Return the response in the expected format.
        throw new Error("Needs to be implemented");
    }
}
```

Implement the methods that require custom handling. Check out the base implementation to check default behavior. If you don't need to override a method, just remove it. You have to define at least `makeRoute`, `preparePayload` and `getResponse`.

If the provider supports multiple tasks that require different implementations, create dedicated subclasses for each task, following the pattern used in the existing providers implementation, e.g. [Together AI provider implementation](https://github.com/huggingface/huggingface.js/blob/main/packages/inference/src/providers/together.ts).

> For text-generation and conversational tasks, one can just inherit from BaseTextGenerationTask and BaseConversationalTask respectively (defined in [providerHelper.ts]((https://github.com/huggingface/huggingface.js/blob/main/packages/inference/src/providers/providerHelper.ts))) and override the methods if needed. Examples can be found in [Cerebras](https://github.com/huggingface/huggingface.js/blob/main/packages/inference/src/providers/cerebras.ts) or [Fireworks](https://github.com/huggingface/huggingface.js/blob/main/packages/inference/src/providers/fireworks.ts) provider implementations.

Reviewer 4: ```suggestion
For text-generation and conversational tasks, you can just inherit from BaseTextGenerationTask and BaseConversationalTask respectively (defined in [providerHelper.ts]((https://github.com/huggingface/huggingface.js/blob/main/packages/inference/src/providers/providerHelper.ts))) and override the methods if needed. Examples can be found in [Cerebras](https://github.com/huggingface/huggingface.js/blob/main/packages/inference/src/providers/cerebras.ts) or [Fireworks](https://github.com/huggingface/huggingface.js/blob/main/packages/inference/src/providers/fireworks.ts) provider implementations.
```

(for consistency)

### 2. Register the provider

Go to [packages/inference/src/lib/getProviderHelper.ts](https://github.com/huggingface/huggingface.js//blob/main/packages/inference/src/lib/getProviderHelper.ts) and add your provider to `PROVIDERS`. Please try to respect alphabetical order. 

### 3. Add tests

Go to [packages/inference/test/InferenceClient.spec.ts](https://github.com/huggingface/huggingface.js/blob/main/packages/inference/test/InferenceClient.spec.ts) and add new tests for each task supported by your provider. 

## 3. Model Mapping API

> Once you've verified the huggingface.js/inference client can call your models successfully, you
> can use our Model Mapping API.
> 
> This API lets a Partner "register" that they support model X, Y or Z on HF.

Reviewer 4: ```suggestion
Congratulations! You now have a JS implementation to successfully make inference calls on your infra! Time to integrate with the Hub!

First step is to use the Model Mapping API to register which HF models are supported. To proceed with this step, we have to enable your account server-side. Make sure you have an organization on the Hub for your enterprise.
```

Reviewer 4: or even put the `First step is to  ...` in a Tip section => they cannot move forward without us

> This enables:
> - The inference widget on corresponding model pages
> - Inference compatibility throughout the HF ecosystem (Python and JS client SDKs for instance), and any downstream tool or library

Reviewer 4: ```suggestion
```

(I feel that it's kinda redundant with intro)

}   
```

### List the whole mapping

```http
GET /api/partners/{provider}/models?status=staging|live
```
This gets all mapping items from the DB. For clarity/DX, the output is grouped by task.

<Tip warning={true}>
This is publicly accessible. It's useful to be transparent by default and it helps debug client SDKs, etc.
</Tip>

Here is an example of response:

```json
{
    "text-to-image": {
        "black-forest-labs/FLUX.1-Canny-dev": {
            "providerId": "black-forest-labs/FLUX.1-canny",
            "status": "live"
        },
        "black-forest-labs/FLUX.1-Depth-dev": {
            "providerId": "black-forest-labs/FLUX.1-depth",
            "status": "live"
        }
    },
    "conversational": {
        "deepseek-ai/DeepSeek-R1": {
            "providerId": "deepseek-ai/DeepSeek-R1",
            "status": "live"
        }
    },
    "text-generation": {
        "meta-llama/Llama-2-70b-hf": {
            "providerId": "meta-llama/Llama-2-70b-hf",
            "status": "live"
        },
        "mistralai/Mixtral-8x7B-v0.1": {
            "providerId": "mistralai/Mixtral-8x7B-v0.1",
            "status": "live"
        }
    }
}
```

> ## 4. Billing

Reviewer 4: Link to billing page somewhere in that section

For routed requests (see figure below), i.e. when users authenticate via HF, our intent is that
our users only pay the standard provider API rates. There's no additional markup from us, we
just pass through the provider costs directly.

<div class="flex justify-center">
    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/inference-providers/types_of_billing.png"/>
</div> 

For LLM providers, a workaround some people use is to extract numbers of input and output
tokens in the responses and multiply by a hardcoded pricing table – this is quite brittle, so we

> are reluctant to do this.

Reviewer 4: Not sure we want to mention this "as-is" in the docs

Author: agreed, I will remove it for now

        """
        return super()._prepare_payload_as_dict(inputs, parameters, mapped_model)

    def _prepare_payload_as_bytes(
        self, inputs: Any, parameters: Dict, mapped_model: str, extra_payload: Optional[Dict]
    ) -> Optional[bytes]:
        """Return the body to use for the request, as bytes.

        Override this method in subclasses for customized body data.
        Only one of `_prepare_payload_as_dict` and `_prepare_payload_as_bytes` should return a value.
        """
        return super()._prepare_payload_as_bytes(inputs, parameters, mapped_model, extra_payload)
```

### 2. Register the Provider
- Go to [src/huggingface_hub/inference/_providers/__init__.py](https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/inference/_providers/__init__.py) and add your provider to `PROVIDER_T` and `PROVIDERS`. Please try to respect alphabetical order. 
- Go to [src/huggingface_hub/inference/_client.py](https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/inference/_client.py) and update docstring in `InferenceClient.__init__` to document your provider.

> ### 3. Add tests

Reviewer 4: Maybe let's not mention that since we are slightly moving away from VCR tests

Author: yes I will remove that for the JS integration as well

    ```python
    _RECOMMENDED_MODELS_FOR_VCR = {
        "your-provider": {
            "task": "model-id",
            ...
        },
        ...
    }
    ```

    b. Set up authentication: To record VCR cassettes, you'll need authentication:
    If you are a member of the provider organization (e.g., Replicate organization: https://huggingface.co/replicate), you can set the HF_INFERENCE_TEST_TOKEN environment variable with your HF token:

    ```bash
    export HF_INFERENCE_TEST_TOKEN="your-hf-token"
    ```
    
    If you're not a member but the provider is officially released on the Hub, you can set the HF_INFERENCE_TEST_TOKEN environment variable as above. If you don't have enough inference credits, we can help you record the VCR cassettes.

    c. Record and commit tests
    Run the tests for your provider:
    ```bash
    pytest tests/test_inference_client.py -k <provider>
    ```

    d. Commit the generated VCR cassettes with your PR.

## FAQ

**Question:** By default, in which order do we list providers in the settings page?

> **Answer:** The default sort is by total number of requests routed by HF over the last 7 days. This order defines which provider will be used in priority by the widget on the model page (but the user's order takes precedence).

Reviewer 4: is that really true? (I haven't followed that)

Author: I'm not sure either, cc @julien-c

Reviewer 2: For now it's random order, but we do want to order by usage eventually indeed

> ## Get in touch!

Reviewer 4: My initial intent was to add this at the very top of the doc page, potentially in a `<Tip>` section to say "Wanna be an Inference Provider on the Hub? Please reach out to us!". Even if we have a proper documentation, it's still best if partners contact us before working too much in the implementation details.
